### PR TITLE
Infer and insert bounds safe interfaces

### PR DESCRIPTION
**Pull request by awruef**
Insert bounds safe interfaces on function declarations, where appropriate. There are a few hacks in here that I would like to re-visit later: 

 - We do a `-1` on SourceLocations to make discrete intervals for the source locations of multi variable declaration statements
 - We do a manual backwards search through the program source text to find the space between the compound statement body of a function definition and the end of the function signature. It would be nice if that information was saved in the AST somewhere.

### CODE REVIEW

File: tools/checked-c-convert/ProgramInfo.cpp [hunk 1]
```diff
@@ -196,6 +199,9 @@ bool PVConstraint::isLt(const ConstraintVariable &Other,
 bool PVConstraint::isEq(const ConstraintVariable &Other,
                         ProgramInfo &Info) const 
 {
+  if (isEmpty() || Other.isEmpty())
```
[the remark this answers is not in the document]
Comment (awruef) [reply tools/checked-c-convert/ProgramInfo.cpp]: I think so? If they're empty, then I'm not sure how they could be different, both philosophically and practically. Also I'm sure some invariant I have unstated somewhere needs this to work. If you want me to track that down, I will, but I'll be sad :(

File: tools/checked-c-convert/ProgramInfo.h [hunk 2]
```diff
@@ -231,14 +238,27 @@ class FunctionVariableConstraint : public ConstraintVariable {
...
   void print(llvm::raw_ostream &O) const;
   void dump() const { print(llvm::errs()); }
   void constrainTo(Constraints &CS, ConstAtom *C, bool checkSkip=false);
   bool anyChanges(Constraints::EnvironmentMap &E);
 
   bool isLt(const ConstraintVariable &other, ProgramInfo &P) const;
   bool isEq(const ConstraintVariable &other, ProgramInfo &P) const;
+  // An FVConstraint is empty if every constraint associated is empty. 
+  bool isEmpty(void) const { 
+    bool allEmpty = false;
```
[the remark this answers is not in the document]
Comment (awruef) [reply tools/checked-c-convert/ProgramInfo.h]: I can fix that.

File: tools/checked-c-convert/ProgramInfo.cpp [hunk 3]
```diff
@@ -485,6 +496,9 @@ bool FVConstraint::isLt(const ConstraintVariable &Other,
 bool FVConstraint::isEq(const ConstraintVariable &Other,
                         ProgramInfo &Info) const 
 {
+  if (isEmpty() || Other.isEmpty())
```
[the remark this answers is not in the document]
Comment (awruef) [reply tools/checked-c-convert/ProgramInfo.cpp]: Same reasoning as above, as well as sadness.

File: tools/checked-c-convert/CheckedCConvert.cpp [hunk 4]
```diff
@@ -105,29 +105,383 @@ ConstraintVariable *getHighest(std::set<ConstraintVariable*> Vs, ProgramInfo &In
...
+
+  for (const auto &D : FD->redecls())
+    if (FunctionDecl *tFD = dyn_cast<FunctionDecl>(D))
+      if (!tFD->isThisDeclarationADefinition())
+        return tFD;
+
+  return FD;
+}
+
+struct DAndReplace
```
[the remark this answers is not in the document]
Comment (awruef) [reply tools/checked-c-convert/CheckedCConvert.cpp]: I'll add a comment, it stands for Decl and replacement.

File: tools/checked-c-convert/ProgramInfo.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (dtarditi) [reply tools/checked-c-convert/ProgramInfo.cpp]: No need to track down an invariant.   I just want to cover the case isEmpty() && Other.isEmpty().

File: tools/checked-c-convert/CheckedCConvert.cpp [hunk 5]
```diff
@@ -504,37 +691,119 @@ bool CastPlacementVisitor::anyTop(std::set<ConstraintVariable*> C) {
...
-        }
-      }
-    }
+  if(Definition == nullptr)
+    return true;
+  assert (Declaration != nullptr);
+
+  // Get constraint variables for the declaration and the definition.
+  // Those constraints should be function constraints. 
+  auto cDecl = cast<FVConstraint>(
```
[the remark this answers is not in the document]
Comment (awruef) [reply tools/checked-c-convert/CheckedCConvert.cpp]: Agreed.

File: tools/checked-c-convert/CheckedCConvert.cpp [hunk 6]
```diff
@@ -504,37 +691,119 @@ bool CastPlacementVisitor::anyTop(std::set<ConstraintVariable*> C) {
...
-        // If they aren't equal, and the constraints in EPC are non-top, 
-        // insert a cast. 
-        if (EPT != PTF && !anyTop(EPC)) {
-          // Insert a cast. 
-          SourceLocation CL = EP->getExprLoc();
-          R.InsertTextBefore(CL, "("+PTF.getAsString()+")");
-        }
-      }
-    }
+  if(Definition == nullptr)
```
[the remark this answers is not in the document]
Comment (awruef) [reply tools/checked-c-convert/CheckedCConvert.cpp]: OK.

File: tools/checked-c-convert/CheckedCConvert.cpp [hunk 7]
```diff
@@ -105,29 +105,383 @@ ConstraintVariable *getHighest(std::set<ConstraintVariable*> Vs, ProgramInfo &In
...
 // to that Decl. The state of the rewrite is contained within R, which
 // is both input and output. R is initialized to point to the 'main'
 // source file for this transformation. toRewrite contains the set of
 // declarations to rewrite. S is passed for source-level information
 // about the current compilation unit.
-void rewrite(Rewriter &R, std::set<DAndReplace> &toRewrite, SourceManager &S,
-             ASTContext &A, std::set<FileID> &Files) {
-  std::set<DAndReplace> skip;
-
+void rewrite( Rewriter              &R,
```
[the remark this answers is not in the document]
Comment (awruef) [reply tools/checked-c-convert/CheckedCConvert.cpp]: I added a comment, will push soon.

File: tools/checked-c-convert/CheckedCConvert.cpp [hunk 8]
```diff
@@ -105,29 +105,383 @@ ConstraintVariable *getHighest(std::set<ConstraintVariable*> Vs, ProgramInfo &In
...
+    // Does the source location of the Decl in lhs overlap at all with
+    // the source location of rhs?
+   
+		SourceRange srLHS = lhs.Declaration->getSourceRange(); 
+    SourceRange srRHS = rhs.Declaration->getSourceRange();
+
+    // Take into account whether or not a FunctionDeclaration specifies 
+    // the "whole" declaration or not. If it does not, it just specifies 
+    // the return position. 
+    if (FunctionDecl *FD = dyn_cast<FunctionDecl>(lhs.Declaration)) {
```
[the remark this answers is not in the document]
Comment (awruef) [reply tools/checked-c-convert/CheckedCConvert.cpp]: I factored it into a helper, will push soon.

File: tools/checked-c-convert/CheckedCConvert.cpp [hunk 9]
```diff
@@ -105,29 +105,383 @@ ConstraintVariable *getHighest(std::set<ConstraintVariable*> Vs, ProgramInfo &In
...
+struct DComp
+{
+  SourceManager &SM;
+  DComp(SourceManager &S) : SM(S) { }
+
+  bool operator()(const DAndReplace lhs, const DAndReplace rhs) const {
+    // Does the source location of the Decl in lhs overlap at all with
+    // the source location of rhs?
+   
+		SourceRange srLHS = lhs.Declaration->getSourceRange();
```
[the remark this answers is not in the document]
Comment (awruef) [reply tools/checked-c-convert/CheckedCConvert.cpp]: Agreed, will fix.

File: tools/checked-c-convert/CheckedCConvert.cpp [hunk 10]
```diff
@@ -105,29 +105,383 @@ ConstraintVariable *getHighest(std::set<ConstraintVariable*> Vs, ProgramInfo &In
...
+      if (tFD->isThisDeclarationADefinition() && tFD->hasBody())
+        return tFD;
+
+  return nullptr;
+}
+
+// Walk the list of declarations and find a declaration that is NOT 
+// a definition and does NOT have a body. 
+FunctionDecl *getDeclaration(FunctionDecl *FD) {
+  for (auto &D : FD->decls())
```
[the remark this answers is not in the document]
Comment (awruef) [reply tools/checked-c-convert/CheckedCConvert.cpp]: Testing this, I think you're right.

File: tools/checked-c-convert/CheckedCConvert.cpp [hunk 11]
```diff
@@ -105,29 +105,383 @@ ConstraintVariable *getHighest(std::set<ConstraintVariable*> Vs, ProgramInfo &In
   return V;
 }
 
-typedef std::pair<Decl*, DeclStmt*> DeclNStmt;
-typedef std::pair<DeclNStmt, std::string> DAndReplace;
+// Walk the list of declarations and find a declaration accompanied by 
+// a definition and a function body. 
+FunctionDecl *getDefinition(FunctionDecl *FD) {
+  for (auto &D : FD->decls())
```
[the remark this answers is not in the document]
Comment (awruef) [reply tools/checked-c-convert/CheckedCConvert.cpp]: Testing this, I think you're right.

**Comment by awruef** [comment]
OK, I think I addressed your feedback.

File: tools/checked-c-convert/ProgramInfo.cpp [hunk 12]
```diff
@@ -188,6 +188,12 @@ bool PVConstraint::liftedOnCVars(const ConstraintVariable &O,
 bool PVConstraint::isLt(const ConstraintVariable &Other, 
                         ProgramInfo &Info) const 
 {
+  if (isEmpty() || Other.isEmpty())
```
Comment (dtarditi) [inline tools/checked-c-convert/ProgramInfo.cpp]: You need to swap the order of the if-statements here.    You should test `A && B` before `A || B`.

File: tools/checked-c-convert/ProgramInfo.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (dtarditi) [reply tools/checked-c-convert/ProgramInfo.cpp]: You need to swap the order of the if-statements here too.

File: tools/checked-c-convert/ProgramInfo.cpp [hunk 13]
```diff
@@ -477,6 +491,12 @@ bool FVConstraint::liftedOnCVars(const ConstraintVariable &Other,
 bool FVConstraint::isLt(const ConstraintVariable &Other,
                         ProgramInfo &Info) const 
 {
+  if (isEmpty() || Other.isEmpty())
```
Comment (dtarditi) [inline tools/checked-c-convert/ProgramInfo.cpp]: You need to swap the if-statements here too.

File: tools/checked-c-convert/ProgramInfo.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (dtarditi) [reply tools/checked-c-convert/ProgramInfo.cpp]: You need to swap the if-statements here.

File: tools/checked-c-convert/CheckedCConvert.cpp [hunk 14]
```diff
@@ -105,221 +105,392 @@ ConstraintVariable *getHighest(std::set<ConstraintVariable*> Vs, ProgramInfo &In
...
-    DeclStmt *Where = DN.second;
-    assert(D != nullptr);
+  return FD;
+}
 
-    if (Verbose) {
-      errs() << "Replacing type of decl:\n";
-      D->dump();
-      errs() << "with " << N.second << "\n";
+struct DAndReplace
```
Comment (dtarditi) [inline tools/checked-c-convert/CheckedCConvert.cpp]: I think you need to add a coment describing the meaning of DAndReplace.

**Comment by awruef** [comment]
If I change the order, everything breaks. I'm investigating why.

**Comment by awruef** [comment]
OK, I figured that out. I think I changed the order in too many places - I added that logic to `isLt` as well as `isEq` and that is not what I want.

**Pull request merged**